services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/service/implement/SecurityGroupServiceImpl.java

```diff
@@ -0,0 +1,242 @@
+
+        securityGroupRepository.addSecurityGroup(oldSecurityGroup);
+        securityGroupJson.setSecurityGroup(oldSecurityGroup);
+
+        LOG.info("Update security group success, securityGroupJson: {}", securityGroupJson);
+
+        return securityGroupJson;
+    }
+
+    private Set<String> getSecurityGroupBindings(String SecurityGroupId) throws Exception {
+        SecurityGroupBindings securityGroupBindings = securityGroupBindingsRepository.getSecurityGroupBindings(SecurityGroupId);
+        if (securityGroupBindings == null || securityGroupBindings.getBindings().size() == 0) {
+            return null;
+        }
+
+        return securityGroupBindings.getBindings();
+    }
+
+    @Override
+    public void deleteSecurityGroup(String securityGroupId) throws Exception {
+        SecurityGroup securityGroup = securityGroupRepository.getSecurityGroup(securityGroupId);
+        if (securityGroup == null) {
+            throw new SecurityGroupNotFound();
+        }
+
+        if (getSecurityGroupBindings(securityGroupId) != null) {
+            throw new SecurityGroupHasBindings();
+        }
+
+        if (securityGroup.getName().equals("default")) {
+            securityGroupRepository.deleteDefaultSecurityGroup(securityGroupId);
+        } else {
+            securityGroupRepository.deleteSecurityGroup(securityGroupId);
+        }
+
+        LOG.info("Delete security group success, securityGroupId: {}", securityGroupId);
+    }
+
+    @Override
+    public SecurityGroupJson getSecurityGroup(String SecurityGroupId) throws Exception {
+        SecurityGroup securityGroup = securityGroupRepository.getSecurityGroup(SecurityGroupId);
+        if (securityGroup == null) {
+            throw new SecurityGroupNotFound();
+        }
+
+        LOG.info("Get security group success, securityGroup: {}", securityGroup);
+
+        return new SecurityGroupJson(securityGroup);
+    }
+
+    @Override
+    public List<SecurityGroupJson> listSecurityGroup() throws Exception {
+        List<SecurityGroupJson> securityGroups = new ArrayList<>();
+
+        Map<String, SecurityGroup> securityGroupMap = securityGroupRepository.getAllSecurityGroups();
+        if (securityGroupMap == null) {
+            return securityGroups;
+        }
+
+        for (Map.Entry<String, SecurityGroup> entry: securityGroupMap.entrySet()) {
+            //Skip the internal default security group
+            if (entry.getKey().equals(entry.getValue().getTenantId())) {
+                continue;
+            }
+
+            SecurityGroupJson securityGroup = new SecurityGroupJson(entry.getValue());
+            securityGroups.add(securityGroup);
+        }
+
+        LOG.info("List security group success");
+
+        return securityGroups;
+    }
+
+    @Override
+    public PortSecurityGroupsJson bindSecurityGroups(PortSecurityGroupsJson portSecurityGroupsJson) throws Exception {
+        securityGroupBindingsRepository.addSecurityGroupBinding(portSecurityGroupsJson);
```

Comment:
For binding, we should also update port manager, right?

Reply:
yes, I will do that after I finish the batch interfaces of security group.